src/basic/chase.c

```diff
@@ -72,7 +72,7 @@ static int log_prohibited_symlink(int fd, ChaseFlags flags) {
                                  strna(n1));
 }
 
-int chaseat(int dir_fd, const char *path, ChaseFlags flags, char **ret_path, int *ret_fd) {
+int chaseat_full(int dir_fd, const char *path, ChaseFlags flags, char **ret_path, char **ret_filename, int *ret_fd) {
```

Comment:
Hmm, I understand the issue, but I wonder if it wouldn't be better solved by adding an optional `root` argument to `path_extract_filename()` which prevents extracting the filename of a directory outside of `root`.

That way we would avoid adding all the `full` variants of the `chase()` functions.

Reply:
- That's super inefficient first mixing the result from `chaseat()` with the (possibly adjusted) root path, and then splitting it again into the root part and sub part in `path_extract_filename()`.
- In `chase()`, the path to `root` directory may not be used as is. So, if we extend `path_extract_filename()`, then we need to sync the logic of root path handling in `chase()` and `path_extract_filename()`. Hence, the maintainability of the code becomes bad...

Reply:
- We have to iterate a few bytes and do one more allocation, that seems OK in terms of efficiency to me. Paths aren't long enough that iterating them is going to be noticeable in profiling
- I don't really follow the second point. If `CHASE_EXTRACT_FILENAME` is used, there's no extra logic needed by the caller to extract the filename. If it is not used, you get the full path and you can just use `path_extract_filename(root, p, &f)` to get the filename. That seems OK to me. If `chaseat()` is used, you can simply do `patch_extract_filename(NULL, p, &f)` to get the filename.